pkg/plugins/golang/options.go

```diff
@@ -0,0 +1,273 @@
+	WebhookVersion string
+
+	// Namespaced is true if the resource should be namespaced.
+	Namespaced bool
+
+	// Flags that define which parts should be scaffolded
+	DoAPI        bool
+	DoController bool
+	DoDefaulting bool
+	DoValidation bool
+	DoConversion bool
+}
+
+// Validate verifies that all the fields have valid values
+func (opts Options) Validate() error {
+	// Check that the required flags did not get a flag as their value
+	// We can safely look for a '-' as the first char as none of the fields accepts it
+	// NOTE: We must do this for all the required flags first or we may output the wrong
+	// error as flags may seem to be missing because Cobra assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupPresent)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionPresent)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindPresent)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the qualified group has a valid DNS1123 subdomain value
+	if err := validation.IsDNS1123Subdomain(opts.QualifiedGroup()); err != nil {
+		return fmt.Errorf("either group or domain is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf("version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	validationErrors := []string{}
+
+	// Require kind to start with an uppercase character
+	if string(opts.Kind[0]) == strings.ToLower(string(opts.Kind[0])) {
+		validationErrors = append(validationErrors, "kind must start with an uppercase character")
+	}
+
+	validationErrors = append(validationErrors, validation.IsDNS1035Label(strings.ToLower(opts.Kind))...)
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("invalid Kind: %#v", validationErrors)
+	}
+
+	// TODO: validate plural strings if provided
+
+	// Ensure apiVersions for k8s types are empty or valid.
+	for typ, apiVersion := range map[string]string{
+		"CRD":     opts.CRDVersion,
+		"Webhook": opts.WebhookVersion,
+	} {
+		switch apiVersion {
+		case "", "v1", "v1beta1":
+		default:
+			return fmt.Errorf("%s version must be one of: v1, v1beta1", typ)
+		}
+	}
+
+	return nil
+}
+
+// QualifiedGroup returns the fully qualified group name with the available information.
+func (opts Options) QualifiedGroup() string {
+	switch "" {
+	case opts.Domain:
+		return opts.Group
+	case opts.Group:
+		return opts.Domain
+	default:
+		return fmt.Sprintf("%s.%s", opts.Group, opts.Domain)
+	}
+}
+
+// GVK returns the GVK identifier of a resource.
+func (opts Options) GVK() resource.GVK {
+	return resource.GVK{
+		Group:   opts.QualifiedGroup(),
+		Version: opts.Version,
+		Kind:    opts.Kind,
+	}
+}
```

Comment:
We have the resouce.GVK and options.GKV. Both are a duplicate code and re returning the same data. It is very confuse. 
IMO Options should use the resource objects instead of duplicate the code. 

And then, I do not agree at all with `Group:   opts.QualifiedGroup(),` as described in https://github.com/kubernetes-sigs/kubebuilder/pull/1911/files#r545354820

Reply:
Both `Options` and `Resources` need to be able to return `GVK` items as they will be used to interact with the config file, and some of these interactions are done before the `Resource` is constructed (e.g., the api scaffolder validate function). The GVK wil be used as the unique key that identifies a resource, for example for functions of the config file like `HasResource` or `GetResource` which will receive a `GVK` as an argument and will return a `bool` or a `Resource` respectively.